### ISSUE TITLE
Carbon support for time-zone-aware doctrine types

### ISSUE THREAD

**Issue by rela589n** [briannesbitt/Carbon#2835]
Hello,

Doctrine has two types, which handle time-zones: `DateTimeTzImmutableType` and `DateTimeTzType`.
For carbon specifically, community has already written couple of types, which implement this:

https://github.com/beberlei/DoctrineExtensions/blob/master/src/Types/CarbonImmutableDateTimeTzType.php
https://github.com/beberlei/DoctrineExtensions/blob/master/src/Types/CarbonDateTimeTzType.php

I would like to propose introduction of two doctrine types to handle time zones:
- carbontz
- carbontz_immutable

**Comment by rela589n** [comment]
Thanks for your response.

I agree that dates should always be stored in a way that allows us to sort/compare consistently. It can either be UTC date or date time with time zone. As a general practice, if we do not save time zone in any way, date must be saved in UTC.

Though, it is not always convenient to operate with UTC dates on the code level. 
Say, on backend we accept date in two fields: datetime, timezone. Then it is converted to universal date time and all operations (incuding storage to database) happen on UTC date. Though, a lot of business logic depend on date time in user timezone, and it is necessary to convert from UTC back to user time zone just for this logic to work correctly. For instance, condition that datetime is after midnight, that date is current month, condition that next form submission is allowed after 7 days (when day 7 comes, user is allowed to submit the form again starting from 12:01am), etc.

Therefore, not at all is such swaying time zones back and forth convenient. For aforementioned cases it would be much more easier for us to operate on dates just in the user time zone. And when it comes to saving into DB, we'd save as `carbontz_immutable`.

**Comment by varshneydevansh** [comment]
I have completed the setup and trying to do this :)

**Comment by varshneydevansh** [comment]
Do we also need common conversion logic to share between the CarbonTzType and CarbonTzImmutableType classes ?

**Comment by kylekatarnls** [comment]
Not necessarily, new types should land here: https://github.com/CarbonPHP/carbon-doctrine-types/tree/main/src/Carbon/Doctrine (doctrine types have been externalized in a separate package since this suggestion was open).

It contains a trait https://github.com/CarbonPHP/carbon-doctrine-types/blob/main/src/Carbon/Doctrine/CarbonTypeConverter.php that help not having duplicate code across type classes.

For instance things such as the maximum precision would be the same for a type with timezone. So we should have write new stuff for that and just re-use the existing trait for everything that works the same.
